upgradejs/depngn
# [feature]: get rid of `npm`/`yarn` commands

This PR removes the use of ~`npm view`~ `npm`/`yarn` commands in favor of reading from `package-lock.json` or `node_modules` directly.

Closes #10 and closes #13 by making them irrelevant.

# Context

The biggest bottleneck in the app right now is the amount of time it takes to run `npm view` for all dependencies. Originally, I was considering using whatever package `npm` uses under the hood in the `view` command to try and optimize, but I realized recently that all of the information we need *should* actually be in the repo it's being run in. This led me to look into just reading the info from file instead of fetching. The execution time difference is wild:

- with `npm view`: ~2s (and this is with a good internet connection on a fast computer)
- reading from filesystem: ~330ms (😱)

(ran these rough benchmarks with `npm` and `yarn` on projects with few dependencies and times were pretty consistent for both -- basically takes less than 1/4 the time of the original implementation)

With this bottleneck out of the way, it opens up a lot more flexibility in how we implement other features, because we can worry a *little* less about performance. It also makes testing easier because we won't have to mock out `fetch` calls to `npm` (plus, tests will run way faster).

# Changes
(**EDIT**: see comment below for additional changes)

- removed `npm view` calls
- replaced them with package manager-specific functions that read the info from the filesystem. `package-lock.json` for `npm` and `node_modules` for `yarn` (by finding `node_modules/{PACKAGE_NAME}/packge.json`)
- added some type info for `package.json` and `package-lock.json`
- renamed `PACKAGE_MANAGER` to `MANAGERS` because I didn't like how it looked 😅 
- added `lockFile: string` to `Manager` objects

I don't believe this would be a breaking change -- it'd just improve performance by a lot.

# QA
- inside the `cli` function (`src/cli/index.ts`), wrap `execWithLog` and `createReport` with `console.time` calls
- switch back and forth between this branch and `main`, running `npm run dev` in each
- compare times
- also, all tests should pass and stuff should work how it's supposed to

```typescript
console.time('depngn');
const compatData = await execWithLog(
  'Fetching engine data',
  async () => await depngn({ version, cwd })
);
await createReport(compatData, version, reporter);
console.timeEnd('depngn');
```
